File: backend/app/rag/text_extractor.py
"""
Text extraction from PDF and DOCX files.
"""
import fitz  # PyMuPDF
from docx import Document as DocxDocument
from pathlib import Path
from dataclasses import dataclass
import logging

logger = logging.getLogger(__name__)


@dataclass
class ExtractedPage:
    text: str
    page_number: int
    source_file: str


class TextExtractor:

    @staticmethod
    async def extract(file_path: str, original_filename: str) -> list[ExtractedPage]:
        path = Path(file_path)
        suffix = path.suffix.lower()

        if suffix == ".pdf":
            return TextExtractor._extract_pdf(file_path, original_filename)
        elif suffix in (".docx", ".doc"):
            return TextExtractor._extract_docx(file_path, original_filename)
        else:
            raise ValueError(f"Unsupported file type: {suffix}")

    @staticmethod
    def _extract_pdf(file_path: str, original_filename: str) -> list[ExtractedPage]:
        pages = []
        try:
            doc = fitz.open(file_path)
            for page_num in range(len(doc)):
                page = doc[page_num]
                text = page.get_text("text")
                if text.strip():
                    pages.append(ExtractedPage(
                        text=text.strip(),
                        page_number=page_num + 1,
                        source_file=original_filename,
                    ))
            doc.close()
            logger.info(f"Extracted {len(pages)} pages from {original_filename}")
        except Exception as e:
            logger.error(f"PDF extraction failed for {original_filename}: {e}")
            raise
        return pages

    @staticmethod
    def _extract_docx(file_path: str, original_filename: str) -> list[ExtractedPage]:
        pages = []
        try:
            doc = DocxDocument(file_path)
            current_text = []
            current_word_count = 0
            page_num = 1

            for para in doc.paragraphs:
                text = para.text.strip()
                if not text:
                    continue
                current_text.append(text)
                current_word_count += len(text.split())

                if current_word_count >= 500:
                    pages.append(ExtractedPage(
                        text="\n".join(current_text),
                        page_number=page_num,
                        source_file=original_filename,
                    ))
                    current_text = []
                    current_word_count = 0
                    page_num += 1

            if current_text:
                pages.append(ExtractedPage(
                    text="\n".join(current_text),
                    page_number=page_num,
                    source_file=original_filename,
                ))

            logger.info(f"Extracted {len(pages)} sections from {original_filename}")
        except Exception as e:
            logger.error(f"DOCX extraction failed for {original_filename}: {e}")
            raise
        return pages

File: backend/src/rag/textExtractor.js
'use strict';
const fs = require('fs');
const path = require('path');
const pdfParse = require('pdf-parse');
const mammoth = require('mammoth');

/**
 * @typedef {{ text: string, pageNumber: number, sourceFile: string }} ExtractedPage
 */

async function extractPdf(filePath, originalFilename) {
  const buffer = fs.readFileSync(filePath);
  const pages = [];

  // pagerender is called once per page by pdf-parse.
  // We collect per-page text here; the callback must return a string.
  await pdfParse(buffer, {
    pagerender: (pageData) => {
      return pageData.getTextContent().then((tc) => {
        const text = tc.items
          .map(item => ('str' in item ? item.str : ''))
          .join(' ')
          .replace(/\s+/g, ' ')
          .trim();

        if (text.length > 0) {
          pages.push({
            text,
            pageNumber: pageData.pageIndex + 1,
            sourceFile: originalFilename,
          });
        }
        return text; // pdf-parse requires a string return value
      });
    },
  });

  // Fallback for PDFs where pagerender doesn't fire (encrypted, image-only, etc.)
  if (pages.length === 0) {
    const data = await pdfParse(buffer);
    const text = data.text.trim();
    if (text) {
      pages.push({ text, pageNumber: 1, sourceFile: originalFilename });
    }
  }

  console.log(`[TextExtractor] Extracted ${pages.length} pages from ${originalFilename}`);
  return pages;
}

async function extractDocx(filePath, originalFilename) {
  const result = await mammoth.extractRawText({ path: filePath });
  const fullText = result.value;

  // Group paragraphs into ~500-word virtual pages
  const paragraphs = fullText.split(/\n+/).map(p => p.trim()).filter(Boolean);
  const pages = [];
  let currentParagraphs = [];
  let currentWordCount = 0;
  let pageNum = 1;

  for (const para of paragraphs) {
    currentParagraphs.push(para);
    currentWordCount += para.split(/\s+/).length;

    if (currentWordCount >= 500) {
      pages.push({
        text: currentParagraphs.join('\n'),
        pageNumber: pageNum,
        sourceFile: originalFilename,
      });
      currentParagraphs = [];
      currentWordCount = 0;
      pageNum++;
    }
  }

  if (currentParagraphs.length > 0) {
    pages.push({
      text: currentParagraphs.join('\n'),
      pageNumber: pageNum,
      sourceFile: originalFilename,
    });
  }

  console.log(`[TextExtractor] Extracted ${pages.length} sections from ${originalFilename}`);
  return pages;
}

async function extract(filePath, originalFilename) {
  const ext = path.extname(filePath).toLowerCase();
  if (ext === '.pdf') return extractPdf(filePath, originalFilename);
  if (ext === '.docx' || ext === '.doc') return extractDocx(filePath, originalFilename);
  throw new Error(`Unsupported file type: ${ext}`);
}

module.exports = { extract };